eval/topdown.go

```diff
@@ -0,0 +1,786 @@
+
+	operands := expr.Terms.([]*opalog.Term)
+	a := operands[1].Value
+	b := operands[2].Value
+
+	return evalEqUnify(ctx, a, b, iter)
+}
+
+func evalEqGround(ctx *TopDownContext, a opalog.Value, b opalog.Value, iter TopDownIterator) error {
+	av, err := valueToInterface(a, ctx)
+	if err != nil {
+		return err
+	}
+	bv, err := valueToInterface(b, ctx)
+	if err != nil {
+		return err
+	}
+	if Compare(av, bv) == 0 {
+		return iter(ctx)
+	}
+	return nil
+}
+
+// evalEqUnify is the top level of the unification implementation.
+//
+// When evaluating equality expressions, OPA tries to unify variables
+// with values or other variables in the expression.
+//
+// The simplest case for unification is an expression of the form "<var> = ???".
+// In this case, the variable is unified/bound to the other side the expression
+// and evaluation continues to the next expression.
+//
+// In cases involving composites, OPA tries to unify elements in the same position
+// of collections. For example, given an expression "[1,2,3] = [1,x,y]", OPA will
+// unify variables x and y with the numbers 2 and 3. This process happens recursively,
+// such that unification can happen on deeply embedded values.
+//
+// In cases involving references, OPA assumes that the references are ground at this stage.
+// As a result, references are just special cases of the normal scalar/composite unification.
+func evalEqUnify(ctx *TopDownContext, a opalog.Value, b opalog.Value, iter TopDownIterator) error {
+
+	// Plug bindings into both terms because this will be called recursively and there may be
+	// new bindings that have been made as part of unification.
+	a = plugValue(a, ctx.Bindings)
+	b = plugValue(b, ctx.Bindings)
+
+	switch a := a.(type) {
+	case opalog.Var:
+		return evalEqUnifyVar(ctx, a, b, iter)
+	case opalog.Object:
+		return evalEqUnifyObject(ctx, a, b, iter)
+	case opalog.Array:
+		return evalEqUnifyArray(ctx, a, b, iter)
+	default:
+		switch b := b.(type) {
+		case opalog.Var:
+			return evalEqUnifyVar(ctx, b, a, iter)
+		case opalog.Array:
+			return evalEqUnifyArray(ctx, b, a, iter)
+		case opalog.Object:
+			return evalEqUnifyObject(ctx, b, a, iter)
+		default:
+			return evalEqGround(ctx, a, b, iter)
+		}
+	}
+
+}
+
+func evalEqUnifyArray(ctx *TopDownContext, a opalog.Array, b opalog.Value, iter TopDownIterator) error {
+	switch b := b.(type) {
+	case opalog.Var:
+		return evalEqUnifyVar(ctx, b, a, iter)
+	case opalog.Ref:
+		return evalEqUnifyArrayRef(ctx, a, b, iter)
+	case opalog.Array:
+		return evalEqUnifyArrays(ctx, a, b, iter)
+	default:
+		return nil
+	}
+}
+
+func evalEqUnifyArrayRef(ctx *TopDownContext, a opalog.Array, b opalog.Ref, iter TopDownIterator) error {
+
+	r, err := ctx.Store.Lookup(b)
+	if err != nil {
+		return err
+	}
+
+	slice, ok := r.([]interface{})
+	if !ok {
+		return nil
+	}
+
+	if len(a) != len(slice) {
+		return nil
+	}
+
+	for i := range a {
+		if isScalar(a[i].Value) {
```

Comment:
Said another way, I would have thought that since b is ground, that we'd look up its value and then call evalEqUnifyArrayArray